### PR TITLE
Implements #81 - webhook event backend

### PR DESCRIPTION
**Pull request by lampwins**
### Fixes: #81 

As discussed in #81, this implements a webhook event backend. Please refer to the conversation in #81 and the docs provided in this PR.

### CODE REVIEW

**Comment by lampwins** [comment]
Hmm, will need figure out a better way to load the webhook cache as for good reason, we should not relay on a DB being available when running `manage.py`...

**Comment by lampwins** [comment]
This was an okay first iteration, but needs some more work.

**Comment by lampwins** [comment]
I have reopened this after refactoring some of the internals and creating a custom signal to handle bulk operations. It is working quite well actually, but there are some things we need to look into.

An empty request is passed into the serializer context to allow `HyperlinkedIdentityField`'s to render. The downside is, it renders relative URLs, leaving out the hostname and protocol. Not sure of a good way around this, as we don't have consistent access to the `request` object.

Other than that, we need to do some extensive testing on the bulk operations side of things.

**Comment by cimnine** [comment]
screen shot 2018-02-02 at 13 20 42

Why is this no valid URL?

**Comment by cimnine** [comment]
> The downside is, it renders relative URLs, leaving out the hostname and protocol. Not sure of a good way around this, as we don't have consistent access to the request object.

As mentioned [in slack][slack-base-url], I suggest to make the base-url configurable as part of the webhook configuration. The netbox instance might be reachable via different hostnames, depending on the calling server.

One particular use case is with Docker: Within the Docker network namespace, netbox might be reachable just as "netbox". But outside the container, it's usually something else.

The same is true for Kubernetes: Within a Pod, the netbox instance might be rechable as "http://netbox", within the whole kubernetes installation as "http://netbox.svc" and outside of it as "https://netbox.my-k8s-cluster.company.com".

[slack-base-url]: https://networktocode.slack.com/archives/C3DQ6MZ0Q/p1517573517000262?thread_ts=1517545406.000244&cid=C3DQ6MZ0Q

**Comment by cimnine** [comment]
[I've also mentioned this in Slack][slack], it would be nice to have a log of outgoing webhooks, so that one can audit and analyze when something weird happens. Which might just be a temporary timeout, 500 error, or whatever.

Something like [Github's webhook tracking][github-webhook] would be probably to much to ask for, but it shows the overall idea very well.



[slack]: https://networktocode.slack.com/archives/C3DQ6MZ0Q/p1517573670000016?thread_ts=1517545406.000244&cid=C3DQ6MZ0Q
[github-webhook]: https://developer.github.com/webhooks/testing/

**Comment by lampwins** [comment]
> Why is this no valid URL?

This has to do with an annoyance in django's builtin `URLField`. Basically it breaks from the RFC and requires a TLD. See https://code.djangoproject.com/ticket/9202 for an explanation on why they have classified it as wontfix. If there is sufficient requirement from users, we can look into doing our own validation via a `RegexField` or similar, but IMHO that's a lot of wheel reinventing for a small gain.

> I've also mentioned this in Slack, it would be nice to have a log of outgoing webhooks, so that one can audit and analyze when something weird happens. Which might just be a temporary timeout, 500 error, or whatever.
> 
> Something like Github's webhook tracking would probably be to much to ask for, but it shows the overall idea very well.

This is for the most part already implemented. Navigate to `/admin/webhook-backend-status/` and you will find status of all rq queues and you can view individual jobs. You may also requeue a failed job. It might not have the fancy UI that github does, but most the same info is present.

**Comment by cimnine** [comment]
> If there is sufficient requirement from users, we can look into doing our own validation via a RegexField or similar, but IMHO that's a lot of wheel reinventing for a small gain.

IMO such webhooks have other targets than "common valid urls". A simple example might be "localhost", and I have mentioned other examples above (i.e. accessing a container by name within a Docker/Kubernetes setup). It should also support plain IPv4s and IPv6s IMO. Or, in other terms, just any valid URL.

A check like https://stackoverflow.com/a/36283503/172132 seems to make sense to me.

**Comment by ngrundler** [comment]
Just wanted to chime in and say that we're really interested in this feature as we're using Salt/NAPALM to do network automation things.  Having Salt events based on changes in NetBox would enable a lot of really cool functionality, so we're hoping to see this merged in soon.

**Comment by lampwins** [comment]
@cimnine the URL issue should be resolved now.

**Comment by lampwins** [comment]
@jeremystretch I have updated this branch for develop-2.4 and it should now be ready for review.

**Comment by jeremystretch** [comment]
@lampwins Excellent, thanks! This is next on my list for v2.4 (finally).

File: netbox/extras/webhooks_worker.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+
+@job('default')
+def process_webhook(webhook, data, model_class, event, signal_received_timestamp):
+    """
+    Make a POST request to the defined Webhook
+    """
+    payload = {
+        'event': event,
+        'signal_received_timestamp': signal_received_timestamp,
```
Comment (jeremystretch) [inline netbox/extras/webhooks_worker.py]: Can we just call this `timestamp`? I realize the signal might not fire at the exact moment an object is saved/deleted, but I think people will assume this is the webhook firing time anyway.

**Review by jeremystretch — COMMENTED** [review]
Really excellent work! Just had a few items I wanted to ask about.

**Pull request merged**